snapcraft/models/project.py

```diff
@@ -1019,6 +1022,20 @@ def _validate_aliases(cls, aliases: list[str]) -> list[str]:
 
         return aliases
 
+    @pydantic.field_validator("common_id")
+    @classmethod
+    def _validate_common_id(cls, common_id: str):
+        common_id_re = re.compile(r"^[a-zA-Z][a-zA-Z0-9_]*(\.[a-zA-Z][a-zA-Z0-9_]*)+$")
```

Comment:
If you make this a `before` validator, you can also put the regex into the pydantic field so it'll be exported to the JSON schema when https://github.com/canonical/snapcraft/issues/5543 gets implemented.